plugin/trino-snowflake/src/main/java/io/trino/plugin/snowflake/SnowflakeClient.java

```diff
@@ -0,0 +1,526 @@
+    private static final DateTimeFormatter SNOWFLAKE_DATE_FORMATTER = DateTimeFormatter.ofPattern("uuuu-MM-dd");
+    private static final DateTimeFormatter SNOWFLAKE_TIMESTAMP_FORMATTER = DateTimeFormatter.ofPattern("u-MM-dd'T'HH:mm:ss.SSSSSSSSS");
+    private static final DateTimeFormatter SNOWFLAKE_TIME_FORMATTER = DateTimeFormatter.ofPattern("HH:mm:ss.SSSSSSSSS");
+    private final AggregateFunctionRewriter<JdbcExpression, ?> aggregateFunctionRewriter;
+
+    private interface WriteMappingFunction
+    {
+        WriteMapping convert(Type type);
+    }
+
+    private interface ColumnMappingFunction
+    {
+        Optional<ColumnMapping> convert(JdbcTypeHandle typeHandle);
+    }
+
+    private static final TimeZone UTC_TZ = TimeZone.getTimeZone(ZoneId.of("UTC"));
+
+    @Inject
+    public SnowflakeClient(
+            BaseJdbcConfig config,
+            ConnectionFactory connectionFactory,
+            QueryBuilder queryBuilder,
+            IdentifierMapping identifierMapping,
+            RemoteQueryModifier remoteQueryModifier)
+    {
+        super("\"", connectionFactory, queryBuilder, config.getJdbcTypesMappedToVarchar(), identifierMapping, remoteQueryModifier, false);
+
+        JdbcTypeHandle bigintTypeHandle = new JdbcTypeHandle(Types.BIGINT, Optional.of("bigint"), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty());
+        ConnectorExpressionRewriter<ParameterizedExpression> connectorExpressionRewriter = JdbcConnectorExpressionRewriterBuilder.newBuilder()
+                .addStandardRules(this::quoted)
+                .build();
+
+        this.aggregateFunctionRewriter = new AggregateFunctionRewriter<>(
+                connectorExpressionRewriter,
+                ImmutableSet.<AggregateFunctionRule<JdbcExpression, ParameterizedExpression>>builder()
+                        .add(new ImplementCountAll(bigintTypeHandle))
+                        .add(new ImplementCount(bigintTypeHandle))
+                        .add(new ImplementMinMax(false))
+                        .add(new ImplementSum(SnowflakeClient::toTypeHandle))
+                        .add(new ImplementAvgFloatingPoint())
+                        .add(new ImplementAvgDecimal())
+                        .build());
+    }
+
+    @Override
+    public Optional<ColumnMapping> toColumnMapping(ConnectorSession session, Connection connection, JdbcTypeHandle typeHandle)
+    {
+        String jdbcTypeName = typeHandle.getJdbcTypeName()
+                .orElseThrow(() -> new TrinoException(JDBC_ERROR, "Type name is missing: " + typeHandle));
+        jdbcTypeName = jdbcTypeName.toLowerCase(Locale.ENGLISH);
+        int type = typeHandle.getJdbcType();
+
+        // Mappings for JDBC column types to internal Trino types
+        final Map<Integer, ColumnMapping> standardColumnMappings = ImmutableMap.<Integer, ColumnMapping>builder()
+                .put(Types.BOOLEAN, StandardColumnMappings.booleanColumnMapping())
```

Comment:
This needs to be updated.

For example, types `INT`, `INTEGER`, `BIGINT`, `SMALLINT`, `TINYINT`, `BYTEINT` are actually synonymous with `NUMBER(38, 0)` in Snowflake ([ref](https://docs.snowflake.com/en/sql-reference/data-types-numeric#int-integer-bigint-smallint-tinyint-byteint)). Their internal "data_type" is always a constant `FIXED` ([ref](https://docs.snowflake.com/en/sql-reference/sql/show-columns#output)). When they're read in via JDBC, their SQL type depends on Snowflake session property `JDBC_TREAT_DECIMAL_AS_INT`. By default its true, so the SQL type is always `Types.BIGINT`. This logic is in Snowflake JDBC code: https://github.com/snowflakedb/snowflake-jdbc/blob/d4ca9f0e99b142fa0c1a3dffe5cc7bde4fdfaf61/src/main/java/net/snowflake/client/jdbc/SnowflakeUtil.java#L177.

So:
1. We actually don't need mappings for `INT`, `INTEGER`, `SMALLINT`, `TINYINT`, etc. as they will never be used.
2. There will be an issue if a user creates an `BIGINT` column in Snowflake and uses it to store a value larger than `Long.MAX_VALUE`.


You can reproduce the example issue:

In Snowflake
```
CREATE TABLE testdb.test1.t_large_bigint (c_bigint BIGINT);

INSERT INTO testdb.test1.t_large_bigint VALUES(99999999999999999999999999999999999998);
```
This query
```
SELECT * FROM testdb.test1.t_large_bigint
```
will succeed in Snowflake but fails in current implementation with an error
```
Query 20240305_052805_00005_6tebh failed: Cannot convert value in the driver from type:FIXED(null,null) to type:Long, value=99999999999999999999999999999999999998.
```

A safer way is to set `JDBC_TREAT_DECIMAL_AS_INT` to false and always read in decimal values.

^The above is an example. It applies to other data types, i.e., `DOUBLE` is used for all floating-point numbers, `VARCHAR` is used for all text types, etc. We will have to reference both Snowflake doc ([here](https://docs.snowflake.com/en/sql-reference/intro-summary-data-types) and [here](https://docs.snowflake.com/en/sql-reference/sql/show-columns#output)) and JDBC code for source-of-truths so that type conversion correctness is ensured: https://github.com/snowflakedb/snowflake-jdbc/blob/d4ca9f0e99b142fa0c1a3dffe5cc7bde4fdfaf61/src/main/java/net/snowflake/client/jdbc/SnowflakeUtil.java#L205-L321


cc @ebyhr @dprophet @yuuteng

Reply:
I will try to fix it when I have time. If you can give me a PR directly, it will be helpful. 😄
@dprophet If you have time you can also take a look on it.

Reply:
Thank you @yuuteng ! Actually I was working on co-authoring a new one but it took longer than I expected. I just commented here in case you folks are working on updates in parallel. I'll see how much I can get it done this week, btw please feel free to work on it on your side as well, I could help comment in that case.

Reply:
@lxynov Lets not fix this right now. Connectors dont need to be perfect on day one. This PR has been going on for over a year. We can do supplimental PR's once this one is merged.

Reply:
@dprophet Skipping this comment is fine. However, please file an issue.

Reply:
Done. https://github.com/trinodb/trino/issues/20977